### PR TITLE
feat: Add script functionality to transformer

### PR DESCRIPTION
**Pull request by Jayclifford345**
This PR adds script support to the transformer tool. This allows creators to create and add scripts to the [sandbox-scripts](https://github.com/grafana/killercoda/tree/scripts/sandbox-scripts) directory. These can be added to the metadata of the source course like so:
```yml
title: Quick start for Tempo
menuTitle: Quick start for Tempo
description: Use Docker to quickly view traces using K-6 and Tempo
weight: 600
killercoda:
  title: Quick start for Tempo
  description: Use Docker to quickly view traces using K-6 and Tempo
  details:
      intro:
         foreground: docker-compose-update.sh
  backend:
    imageid: ubuntu
```
When the transformer runs it goes through the following steps:
1. Check the metadata for included scripts. To be valid they must be nested under `details` and we currently only support foreground and background scripts for `intro` and `finish`. I have not included step scripts as this can get a little messy.
2. It will then look for the scripts within the `sandbox-scripts` directory. If they do exist the transformer will fail.
3. If they do exist they are cloned to the tutorial directory. Note we need to do this due to the fact Killercoda currently does not support pathing to scripts out with the tutorial directory.
4. Clean up any other shell scripts in the tutorial directory which are not specified within the metadata. This makes sure old scripts are not being left causing unneeded space and complexity to the tutorial.

### CODE REVIEW

**Comment by Jayclifford345** [comment]
@jdbaldry Thats the functions removed. Ready for your next review!

File: docs/transformer.md [hunk 1]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
-         }
-      },
-      "backend": {
-         "imageid": "ubuntu"
-      }
-   }
-   ```
-   Refer to the [grafana-basics](../grafana/grafana-basics/index.json) tutorial for an example.
\ No newline at end of file
+## Adding foreground and background scripts (experimental)
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
## Add foreground and background scripts
```

File: docs/transformer.md [hunk 2]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+
+> Note: It is recommended to use `foreground` scripts rather than `background` scripts. This allows the user to see the output of the script. 
+
+Foreground and background scripts are shell scripts that run during the introduction and finish pages of a tutorial. The scripts are useful for setting up the environment for the tutorial and cleaning up after the tutorial. Scripts are another asset that needs to be maintained, so use them sparingly. A good example of using a script is to update the `docker compose` package before running the tutorial.
```
Comment (jdbaldry) [inline docs/transformer.md]: I think documenting the difference after introducing them is 
more valuable than a note about the difference before a reader knows what they are

```suggestion
Foreground and background scripts are shell scripts that run during the introduction and finish pages of a tutorial.
The scripts are useful for setting up the environment for the tutorial and cleaning up after the tutorial.

Scripts are another asset that needs to be maintained, so use them sparingly.
A good example of using a script is to update the Docker Compose package before running the tutorial.

Use foreground scripts when you want to the user to see the script output.
Use background scripts when you want to hide the output of the script.
```

File: docs/transformer.md [hunk 3]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+
+### Create your script
+
+Since these scripts are primarily used for preparing the interactive environment, they are stored within the [`sandbox-scripts`](/sandbox-scripts/) directory in the Killercoda repository. Make sure to create your script in this directory and name with a unique name that reflects the purpose of the script. Note if a script already exists that performs the same function, you can reuse it.
+
+Here is an example of a script that updates the `docker-compose` package:
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
Here is an example of a script that updates the Docker Compose package:
```

File: docs/transformer.md [hunk 4]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
\ No newline at end of file
+## Adding foreground and background scripts (experimental)
+
+> Note: It is recommended to use `foreground` scripts rather than `background` scripts. This allows the user to see the output of the script. 
+
+Foreground and background scripts are shell scripts that run during the introduction and finish pages of a tutorial. The scripts are useful for setting up the environment for the tutorial and cleaning up after the tutorial. Scripts are another asset that needs to be maintained, so use them sparingly. A good example of using a script is to update the `docker compose` package before running the tutorial.
+
+### Create your script
+
+Since these scripts are primarily used for preparing the interactive environment, they are stored within the [`sandbox-scripts`](/sandbox-scripts/) directory in the Killercoda repository. Make sure to create your script in this directory and name with a unique name that reflects the purpose of the script. Note if a script already exists that performs the same function, you can reuse it.
```
Comment (jdbaldry) [inline docs/transformer.md:299]: Is this still correct?

File: docs/transformer.md [hunk 5]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+sudo chmod a+r /etc/apt/keyrings/docker.asc
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+sudo apt-get update
+sudo apt-get install -y docker-compose-plugin && clear && echo "Setup complete. You may now begin the tutorial."
+```
+
+Note that you should add a message at the end of the script to inform the user that the setup is complete.
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
> [!TIP]
> Add a message at the end of the script to inform the user that the setup is complete.
```

File: tempo/quick-start/docker-compose-update.sh [hunk 6]
```diff
@@ -0,0 +1,9 @@
+sudo install -m 0755 -d /etc/apt/keyrings
+sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+sudo chmod a+r /etc/apt/keyrings/docker.asc
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+sudo apt-get update
+sudo apt-get install -y docker-compose-plugin && clear && echo "Setup complete. You may now begin the tutorial."
```
Comment (jdbaldry) [inline tempo/quick-start/docker-compose-update.sh]: May need some testing to double check the behavior but here's a ShellCheck linted POSIX shell script.

By the way, do we know which user this runs as? Do we need all the `sudo` commands? I guess they don't hurt.

```suggestion
#!/bin/sh

set -euf
# shellcheck disable=SC3040
(set -o pipefail 2> /dev/null) && set -o pipefail


sudo install -m 0755 -d /etc/apt/keyrings
sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
sudo chmod a+r /etc/apt/keyrings/docker.asc

ARCH="$(dpkg --print-architecture)"
VERSION_CODENAME="$(source /etc/os-release && echo "${VERSION_CODENAME}")"
readonly ARCH VERSION_CODENAME

printf 'deb [arch=%s signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu %s stable' "${ARCH}" "${VERSION_CODENAME}" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null

sudo apt-get update
sudo apt-get install -y docker-compose-plugin

echo "Setup complete. You may now begin the tutorial."
```

File: tempo/quick-start/intro.md [hunk 7]
```diff
@@ -8,3 +8,5 @@ The Tempo examples running with Docker using docker-compose include a version of
 
 This quick start guide uses the `local`{{copy}} example running Tempo as a single binary (monolithic). Any data is stored locally in the `tempo-data`{{copy}} folder.
 To learn more, read the [local storage example README](https://github.com/grafana/tempo/blob/main/example/docker-compose/local).
+
+> **Note**: A setup script is running in the background to install the necessary dependencies. This should take no longer than 30 seconds. Your instance will be ready to use once you `Setup complete. You may now begin the tutorial`.
```
Comment (jdbaldry) [inline tempo/quick-start/intro.md]:
```suggestion
> **Note**:
> A setup script is running to install the necessary dependencies.
> This should take no longer than 30 seconds. 
> Your instance will be ready to use after you see `Setup complete. You may now begin the tutorial`.
```

**Review by jdbaldry — COMMENTED** [review]
Code changes LGTM.

File: docs/transformer.md (hunk 4 above)
Comment (jdbaldry) [inline docs/transformer.md]: I think this needs updating

File: docs/transformer.md [hunk 8]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+sudo install -m 0755 -d /etc/apt/keyrings
+sudo curl -fsSL https://download.docker.com/linux/ubuntu/gpg -o /etc/apt/keyrings/docker.asc
+sudo chmod a+r /etc/apt/keyrings/docker.asc
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.asc] https://download.docker.com/linux/ubuntu \
+  $(. /etc/os-release && echo "$VERSION_CODENAME") stable" | \
+  sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+sudo apt-get update
+sudo apt-get install -y docker-compose-plugin && clear && echo "Setup complete. You may now begin the tutorial."
+```
```
Comment (jdbaldry) [inline docs/transformer.md]: If we update the Tempo script, I think we should update this copy too.

File: docs/transformer.md [hunk 9]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+
+Note that you should add a message at the end of the script to inform the user that the setup is complete.
+
+### Add the script to the tutorial
+
+To add the script to the tutorial, you need to add the script to the `killercoda` metadata in the source file. Note that we currently only support foreground and background scripts for the introduction and finish pages. These scripts are specified in the `killercoda.details` field in the source file. Lets look at an example:
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
To add the script to the tutorial, you need to add the script to the `killercoda` metadata in the source file. 

> [!NOTE]
> The transformer tool only supports foreground and background scripts for the introduction and finish pages.

The following example sets the foreground script for the introduction page to be `docker-compose-update.sh`:
```

File: docs/transformer.md [hunk 10]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+  title: Quick start for Tempo
+  description: Use Docker to quickly view traces using K-6 and Tempo
+  details:
+      intro:
+         foreground: docker-compose-update.sh
+  backend:
+    imageid: ubuntu
+```
+
+In the example above, we have added a foreground script to the introduction page. The script is named `docker-compose-update.sh` and is located in the `sandbox-scripts` directory. The script will run when the introduction page is loaded.
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
```

File: docs/transformer.md [hunk 11]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+
+Another example is adding a background script to the finish page:
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
The following example sets the foreground script for the introduction page to be `docker-compose-cleanup.sh`:
```

File: docs/transformer.md [hunk 12]
```diff
@@ -288,63 +288,79 @@ To generate a tutorial:
...
+Here are a few common facts about using foreground and background scripts:
+
+* Any scripts which are not specified in the `killercoda.details` field but exist in the course directory will be **deleted**. This helps us to clean up old scripts that are no longer in use.
+
+* If you want to use the same script for multiple tutorials, you can specify the same script in multiple source files. The script will be copied to the respective tutorial directories.
+
+* The transformer will fail if the script specified in the `killercoda.details` field does not exist in the `sandbox-scripts` directory.
+
+* We currently only support foreground and background scripts for the introduction and finish pages. If you require scripts for other pages, please reach out to us via an issue.
+
```
Comment (jdbaldry) [inline docs/transformer.md]:
```suggestion
```

File: docs/transformer.md (hunk 4 above)
Comment (jdbaldry) [reply docs/transformer.md]:
```suggestion
Create your script in the course directory sure with a name that reflects the purpose of the script.
```

**Review by jdbaldry — APPROVED** [review]

**Pull request merged**